### PR TITLE
chore: Installing packages webpack@4 babel@7 rollup...

### PR DESCRIPTION
**Pull request by xiaoyuhen**
**Summary**
- upgrade webpack dependencies
  - upgrade webpack@4.28.4
    - use mini-css-extract-plugin instead extract-text-webpack-plugin
    - remove ```NoErrorsPlugin```,use the optimization.noEmitOnErrors to skip the emitting phase whenever there are errors while compiling
    - webpack.optimize.DedupePlugin isn't needed anymore. Remove it from the configuration
    - The OccurrenceOrderPlugin is now enabled by default and has been renamed. Remove ```OccurenceOrderPlugin``` from the configuration
    - webpack.optimize.UglifyJsPlugin has been removed, use config.optimization.minimize instead
    - change ```loaders ``` to ```rules ```
    - use __dirname to get the current directory and then append ./build
    - remove '' from extensions array, see https://github.com/webpack/webpack/issues/3043
  - upgrade clean-webpack-plugin@1.0.0
  - upgrade copy-webpack-plugin@4.6.0
  - upgrade webpack-dev-server@3.1.14
  - install webpack-cli@3.2.1
  - upgrade sass-loader@7.1.0
  - upgrade node-sass@4.11.0
  - upgrade css-loader@2.1.0
  - upgrade file-loader@3.0.1
  - upgrade json-loader@0.5.7
  - upgrade url-loader@1.1.2
  - upgrade static-site-generator-webpack-plugin@3.4.2
  - upgrade raw-loader@1.0.0
    - specify raw-loader instead of raw

- upgrade babel@7
  - babel-preset-es2015、babel-preset-stage-0 ... -> @babel/preset-env,see https://babeljs.io/docs/en/env
  - babel-preset-react ->  @babel/preset-react
  - babel-plugin-transform-object-rest-spread -> @babel/plugin-proposal-object-rest-spread
  - babel-cli -> @babel/cli
  - upgrade babel-loader@8.0.5
  - upgrade @babel/core@7.2.2
  - install @babel/plugin-transform-runtime@7.2.0
  - upgrade babel-eslint@10.0.1
  - update babel jest config, see [babel-jest-example](https://github.com/facebook/jest/tree/54f4d4ebd3d1a11d65962169f493ce41efdd784f/examples/babel-7)
    - install babel-jest、regenerator-runtime、babel-core@^7.0.0-bridge.0
  - update babel config
  - modify webpack babel-loader config

- upgrade rollup
  - copy rollup config from ```react-reactstrap```

- other
  - remove react-test-renderer, because we don't use it
  - upgrade reactstrap@7.0.2
  - upgrade eslint@5.0.0
  - upgrde ejs@2.5.9
  - upgrade conventional-changelog-cli@1.3.22
  - fix ```compound selectors may longer be extended```,see http://bit.ly/ExtendCompound

**test plan**

- [x] ```npm run test``` is running normally
- [x] ```npm run build``` is running normally
- [x] ```npm run build-docs``` is running normally
- [x] ```npm start``` is running normally
- [x]  ```sh ./scripts/docs``` is running normally

### CODE REVIEW

**Comment by JimLiu** [comment]
Please take a look with https://github.com/reactstrap/reactstrap
Video-React uses the same scaffold as reactstrap, and it had upgraded webpack4, so we just need to follow what reactstrap does.

**Comment by xiaoyuhen** [comment]
@JimLiu This is a very good suggestion, I will take a look at ```reactstrap``` first.

**Comment by llaver** [comment]
Very excited for this. Tried to do it myself a little while back and didn't even get close.

**Comment by mondaychen** [comment]
Thank you for your hard work @xiaoyuhen. Will take a look today.

File: package.json [hunk 1]
```diff
@@ -65,6 +66,7 @@
   },
   "dependencies": {
     "classnames": "^2.2.3",
+    "core-js": "^2.6.3",
```
Comment (mondaychen) [inline package.json]: Why is this needed? A required peer dependency?

File: rollup.config.js [hunk 2]
```diff
@@ -63,9 +79,9 @@ function baseUmdConfig(minified) {
 const libConfig = baseConfig();
 // Do not include any of the dependencies
 libConfig.external = peerDependencies.concat(dependencies);
-libConfig.targets = [
-  { dest: 'dist/video-react.cjs.js', format: 'cjs' },
-  { dest: 'dist/video-react.es.js', format: 'es' },
+libConfig.output = [
+  { sourcemap: true, name: 'Reactstrap', file: 'dist/video-react.cjs.js', format: 'cjs' },
```
Comment (mondaychen) [inline rollup.config.js]: This `name` should be something like "video-react", right?

File: styles/scss/components/progress.scss [hunk 3]
```diff
@@ -53,8 +53,7 @@
 .video-react .video-react-mouse-display {
   @extend .video-react-icon;
 
-  &:before {
-    @extend .video-react-icon-circle:before;
+  .video-react-icon-circle:before, &:before {
```
Comment (mondaychen) [inline styles/scss/components/progress.scss]: I'm confused.

Why did it need to extend this style when it's hidden? @JimLiu 

Why did you make this change? @xiaoyuhen

File: rollup.config.js [hunk 4]
```diff
@@ -10,26 +10,42 @@ const packageJson = require('./package.json');
 const peerDependencies = Object.keys(packageJson.peerDependencies);
 const dependencies = Object.keys(packageJson.dependencies);
 
+function globals() {
```
Comment (mondaychen) [inline rollup.config.js]: Now looks like we have two places defining "globals": here, and `umdFullConfig.globals`. Let's try to only use one.

File: styles/scss/components/progress.scss (hunk 3 above)
Comment (JimLiu) [reply styles/scss/components/progress.scss]: should not change the css file

File: rollup.config.js [hunk 5]
```diff
@@ -63,9 +79,9 @@ function baseUmdConfig(minified) {
 const libConfig = baseConfig();
 // Do not include any of the dependencies
 libConfig.external = peerDependencies.concat(dependencies);
-libConfig.targets = [
-  { dest: 'dist/video-react.cjs.js', format: 'cjs' },
-  { dest: 'dist/video-react.es.js', format: 'es' },
+libConfig.output = [
+  { sourcemap: true, name: 'Reactstrap', file: 'dist/video-react.cjs.js', format: 'cjs' },
+  { sourcemap: true, name: 'Reactstrap', file: 'dist/video-react.es.js', format: 'es' },
```
Comment (JimLiu) [inline rollup.config.js]: change `Reactstrap` to `video-react`

**Comment by JimLiu** [comment]
@xiaoyuhen 
Thanks for your great work on this change. it looks good to me overall, just need to make some small changes based on the comment.

And please test `sh ./scripts/docs` to see if it can generate the static website.

Thanks again.

**Comment by xiaoyuhen** [comment]
@JimLiu 

Github provides Rabase and Squash functionality, maybe we can use it instead of manually merging MR


I have added ```sh ./scripts/docs``` to the test plan, I will test later

File: styles/scss/components/progress.scss (hunk 3 above)
Comment (mondaychen) [reply styles/scss/components/progress.scss]: Looks like this whole `.video-react .video-react-mouse-display` block is not really used. I tried to remove it and `MouseTimeDisplay` component still looks good. Let's just get rid of it.

**Comment by mondaychen** [comment]
@xiaoyuhen after #204 is merged, there's gonna be some conflicts in package.json. Please help resolve it. If you need any help from me, don't hesitate to ask! Thank you again for your great work!

File: styles/scss/components/progress.scss (hunk 3 above)
Comment (xiaoyuhen) [reply styles/scss/components/progress.scss]: already remove this style

**Comment by xiaoyuhen** [comment]
@JimLiu @mondaychen 

All request changes have been fixed, for this pr, if you have any other questions or confusion, please let me know

File: rollup.config.js (hunk 4 above)
Comment (mondaychen) [inline rollup.config.js]: I think we can remove `globals` from `baseUmdConfig` now. It's not used anymore

Comment (mondaychen) [reply rollup.config.js]: I got this warning when running `npm run build`:


```
(!) You have passed an unrecognized option
Unknown input option: globals. Allowed options: acorn, acornInjectPlugins, cache, chunkGroupingSize, context, experimentalCacheExpiry, experimentalOptimizeChunks, experimentalTopLevelAwait, external, inlineDynamicImports, input, manualChunks, moduleContext, onwarn, perf, plugins, preserveModules, preserveSymlinks, shimMissingExports, treeshake, watch
```

Removing `globals` from the config will also fix this.

**Comment by mondaychen** [comment]
Can't wait any longer for this. 😃 

Let me just merge this first and fix the `globals` thing and `rollup-plugin-babel-minify` thing myself. They are just warnings anyway.

This is awesome job @xiaoyuhen
